src/core/QCheck.ml

```diff
@@ -395,8 +395,45 @@ module Shrink = struct
         done
 
   let list ?shrink l yield =
```

Comment:
looks good. I'll refactor the style a bit but the idea is good in any case (similar to integer shrinking I think).

Reply:
Great. Besides the repetitive testing for empty list `[]` the code should mirror the array-based strategy from earlier, as confirmed by matching the exact number of successful shrink steps between the two.